File: janggan/compare.py
"""5단계 — 두 모델 생성물의 구조 유효성 비교 (연구 ④).

각 모델에서 온도별 N개 생성 → 디코딩 → 유효성 측정:
  parse   토큰이 문법적으로 파싱되나
  cycle   총 길이가 학습 장단 길이(9/10/12)에 드나        ← 양쪽 공통 지표
  shape   박 구조가 학습 장단 shape 에 드나                 ← v2 전용(v1은 박 경계 부재)
  novel   학습셋에 없는 새 시퀀스인가                        ← 암기 아닌 생성 확인

핵심 가설(3단계서 수정): v1 은 박 경계를 토큰에 못 담아 cycle 유효율이 더 낮을 것.
"""

from __future__ import annotations

import json
from pathlib import Path

import torch

from dataset import BOS, EOS, PAD, ENCODERS, build
from model import TinyGPT

RUNS = Path(__file__).with_name("runs")
DATA = Path(__file__).with_name("data") / "patterns.json"
GU = {"deong", "kung", "deok", "gideok", "roll"}
N, TEMPS, SEED = 200, [1.0, 1.5, 2.0], 7


# ── 안전 파서 (생성물은 malformed 일 수 있음) ────────────────────────────────

def v1_parse(toks):
    if len(toks) == 0 or len(toks) % 2 != 0:
        return None
    flat = []
    for a in range(0, len(toks), 2):
        sym, dur = toks[a], toks[a + 1]
        if sym not in GU and sym != "rest":
            return None
        if not (dur.startswith("d") and dur[1:].isdigit() and int(dur[1:]) >= 1):
            return None
        flat.append(None if sym == "rest" else sym)
        flat += [None] * (int(dur[1:]) - 1)
    return flat


def v2_parse(toks):
    if not toks or toks[0][0] != "|":
        return None
    grid, beat, i = [], None, 0
    while i < len(toks):
        t = toks[i]
        if t[0] == "|":
            if beat is not None:
                grid.append(beat)
            if not t[1:].isdigit():
                return None
            beat = [None] * int(t[1:])
            i += 1
        elif t[0] == "p":
            if not t[1:].isdigit():
                return None
            pos = int(t[1:])
            if pos >= len(beat) or i + 1 >= len(toks) or toks[i + 1] not in GU:
                return None
            if beat[pos] is not None:
                return None
            beat[pos] = toks[i + 1]
            i += 2
        else:
            return None
    grid.append(beat)
    return grid


# ── 생성·측정 ────────────────────────────────────────────────────────────────

def load(scheme, dev):
    ck = torch.load(RUNS / f"{scheme}.pt", map_location=dev, weights_only=False)
    cfg = ck["cfg"]
    m = TinyGPT(len(ck["vocab"]), ck["maxlen"], ck["ids"]["pad"],
                cfg["d"], cfg["h"], cfg["layers"], cfg["dropout"]).to(dev)
    m.load_state_dict(ck["state"])
    m.eval()
    stoi = {t: i for i, t in enumerate(ck["vocab"])}
    return m, ck["vocab"], stoi, ck["ids"]


def measure(scheme, patterns, dev, valid_totals, valid_shapes):
    m, vocab, stoi, ids = load(scheme, dev)
    parse = v1_parse if scheme == "v1" else v2_parse
    train_set = {tuple(ENCODERS[scheme](p["grid"])) for p in patterns}

    rows = []
    for temp in TEMPS:
        torch.manual_seed(SEED)
        ok = cyc = shp = nov = 0
        samples = []
        for _ in range(N):
            idx = torch.tensor([[ids["bos"]]], device=dev)
            out = m.generate(idx, max_new=m.block_size, eos_id=ids["eos"], temperature=temp)
            toks = [vocab[i] for i in out[0].tolist()]
            toks = [t for t in toks if t not in (BOS, EOS, PAD)]
            parsed = parse(toks)
            if parsed is None:
                continue
            ok += 1
            total = len(parsed) if scheme == "v1" else sum(len(b) for b in parsed)
            if total in valid_totals:
                cyc += 1
            if scheme == "v2" and tuple(len(b) for b in parsed) in valid_shapes:
                shp += 1
            if tuple(toks) not in train_set:
                nov += 1
            samples.append(toks)
        rows.append({
            "temp": temp,
            "parse_%": round(100 * ok / N, 1),
            "cycle_%": round(100 * cyc / N, 1),
            "shape_%": round(100 * shp / N, 1) if scheme == "v2" else None,
            "novel_%": round(100 * nov / N, 1),
        })
    return rows, samples


if __name__ == "__main__":
    dev = torch.device("mps") if torch.backends.mps.is_available() else torch.device("cpu")
    patterns = json.loads(DATA.read_text(encoding="utf-8"))["patterns"]
    valid_totals = {sum(len(b) for b in p["grid"]) for p in patterns}
    valid_shapes = {tuple(len(b) for b in p["grid"]) for p in patterns}
    print(f"device={dev}  N={N}/temp  valid_totals={sorted(valid_totals)}  "
          f"valid_shapes={sorted(valid_shapes)}\n")

    report = {}
    for scheme in ("v1", "v2"):
        rows, samples = measure(scheme, patterns, dev, valid_totals, valid_shapes)
        report[scheme] = rows
        print(f"[{scheme}]")
        for r in rows:
            shp = f" shape={r['shape_%']:5}" if r["shape_%"] is not None else ""
            print(f"  temp {r['temp']}  parse={r['parse_%']:5}%  "
                  f"cycle={r['cycle_%']:5}%{shp}  novel={r['novel_%']:5}%")
        print(f"  예시 생성(temp {TEMPS[-1]}): {' '.join(samples[0]) if samples else '—'}\n")

    (RUNS / "compare.json").write_text(json.dumps(report, ensure_ascii=False, indent=2))
    print("saved → runs/compare.json")

File: janggan-ts/src/compare.ts
// 5단계 — 생성물 구조 유효성 비교 — 파이썬 compare.py 이식.
//   parse / cycle(공통) / shape(v2전용) / novel.

import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { BOS, EOS, PAD, GU_SET, type Cell, type Grid } from "./tokens.ts";
import { ENCODERS, loadPatterns } from "./encode.ts";
import { build } from "./dataset.ts";
import { TinyGPT } from "./model.ts";

const N = 200;
const TEMPS = [1.0, 1.5, 2.0];
const SEED = 7;

const HERE = dirname(fileURLToPath(import.meta.url));
const RUNS = join(HERE, "..", "runs");

// ── 안전 파서 (생성물은 malformed 일 수 있음) ────────────────────────────────
function v1Parse(toks: string[]): Cell[] | null {
  if (toks.length === 0 || toks.length % 2 !== 0) return null;
  const flat: Cell[] = [];
  for (let a = 0; a < toks.length; a += 2) {
    const sym = toks[a];
    const dur = toks[a + 1];
    if (!GU_SET.has(sym) && sym !== "rest") return null;
    if (!/^d\d+$/.test(dur) || Number(dur.slice(1)) < 1) return null;
    flat.push(sym === "rest" ? null : (sym as Cell));
    for (let r = 0; r < Number(dur.slice(1)) - 1; r++) flat.push(null);
  }
  return flat;
}

function v2Parse(toks: string[]): Grid | null {
  if (toks.length === 0 || toks[0][0] !== "|") return null;
  const grid: Grid = [];
  let beat: Cell[] | null = null;
  let i = 0;
  while (i < toks.length) {
    const t = toks[i];
    if (t[0] === "|") {
      if (beat !== null) grid.push(beat);
      if (!/^\d+$/.test(t.slice(1))) return null;
      beat = new Array(Number(t.slice(1))).fill(null);
      i += 1;
    } else if (t[0] === "p") {
      if (!/^\d+$/.test(t.slice(1))) return null;
      const pos = Number(t.slice(1));
      if (beat === null || pos >= beat.length || i + 1 >= toks.length || !GU_SET.has(toks[i + 1])) return null;
      if (beat[pos] !== null) return null;
      beat[pos] = toks[i + 1] as Cell;
      i += 2;
    } else {
      return null;
    }
  }
  if (beat !== null) grid.push(beat);
  return grid;
}

interface Row { temp: number; parse: number; cycle: number; shape: number | null; novel: number }

function measure(scheme: string, validTotals: Set<number>, validShapes: Set<string>) {
  const patterns = loadPatterns();
  const { vocab, ids } = build(patterns, scheme);
  const model = TinyGPT.fromJSON(readFileSync(join(RUNS, `${scheme}.model.json`), "utf-8"));
  const parse = scheme === "v1" ? v1Parse : v2Parse;
  const trainSet = new Set(patterns.map((p) => ENCODERS[scheme](p.grid).join(" ")));

  const rows: Row[] = [];
  let sample = "";
  for (const temp of TEMPS) {
    let ok = 0, cyc = 0, shp = 0, nov = 0;
    for (let n = 0; n < N; n++) {
      const out = model.generate(ids.bos, ids.eos, model.cfg.block, temp, SEED * 1000 + n * 13);
      const toks = out.map((i) => vocab[i]).filter((t) => t !== BOS && t !== EOS && t !== PAD);
      const parsed = parse(toks);
      if (parsed === null) continue;
      ok++;
      const total = scheme === "v1"
        ? (parsed as Cell[]).length
        : (parsed as Grid).reduce((s, b) => s + b.length, 0);
      if (validTotals.has(total)) cyc++;
      if (scheme === "v2" && validShapes.has((parsed as Grid).map((b) => b.length).join(","))) shp++;
      if (!trainSet.has(toks.join(" "))) nov++;
      if (temp === TEMPS[TEMPS.length - 1] && !sample) sample = toks.join(" ");
    }
    rows.push({
      temp,
      parse: +(100 * ok / N).toFixed(1),
      cycle: +(100 * cyc / N).toFixed(1),
      shape: scheme === "v2" ? +(100 * shp / N).toFixed(1) : null,
      novel: +(100 * nov / N).toFixed(1),
    });
  }
  return { rows, sample };
}

const patterns = loadPatterns();
const validTotals = new Set(patterns.map((p) => p.grid.reduce((s, b) => s + b.length, 0)));
const validShapes = new Set(patterns.map((p) => p.grid.map((b) => b.length).join(",")));
console.log(`N=${N}/temp  valid_totals=${[...validTotals].sort()}  valid_shapes=${[...validShapes].sort()}\n`);

const report: Record<string, Row[]> = {};
for (const scheme of ["v1", "v2"]) {
  const { rows, sample } = measure(scheme, validTotals, validShapes);
  report[scheme] = rows;
  console.log(`[${scheme}]`);
  for (const r of rows) {
    const shp = r.shape !== null ? ` shape=${r.shape.toString().padStart(5)}` : "";
    console.log(`  temp ${r.temp}  parse=${r.parse.toString().padStart(5)}%  cycle=${r.cycle.toString().padStart(5)}%${shp}  novel=${r.novel.toString().padStart(5)}%`);
  }
  console.log(`  예시 생성(temp ${TEMPS[TEMPS.length - 1]}): ${sample || "—"}\n`);
}

writeFileSync(join(RUNS, "compare.json"), JSON.stringify(report, null, 2));
console.log("saved → runs/compare.json");
